server/src/main/java/io/unitycatalog/server/auth/decorator/UnityAccessDecorator.java

```diff
@@ -0,0 +1,402 @@
+    // we want to authorize against.
+
+    Map<ResourceType, Object> resourceKeys = new HashMap<>();
+
+    // Split up the locators by type, because we have to extract the value from the request
+    // different ways for different types
+
+    List<KeyLocator> systemLocators = locators.stream().filter(l -> l.getSource().equals(SYSTEM)).toList();
+    List<KeyLocator> paramLocators = locators.stream().filter(l -> l.getSource().equals(PARAM)).toList();
+    List<KeyLocator> payloadLocators = locators.stream().filter(l -> l.getSource().equals(PAYLOAD)).toList();
+
+    // Add system-type keys, just metastore for now.
+    systemLocators.forEach(l -> resourceKeys.put(l.getType(), "metastore"));
+
+    // Extract the query/path parameter values just by grabbing them from the request
+    paramLocators.forEach(l -> {
+      String value = ctx.pathParam(l.getKey()) != null ? ctx.pathParam(l.getKey()) : ctx.queryParam(l.getKey());
+      resourceKeys.put(l.getType(), value);
+    });
+
+    if (payloadLocators.isEmpty()) {
+      // If we don't have any PAYLOAD locators, we're ready to evaluate the authorization and allow or deny
+      // the request.
+      LOGGER.debug("Checking authorization before method.");
+      checkAuthorization(principal, expression, resourceKeys);
+
+      return delegate.serve(ctx, req);
+    } else {
+      // Since we have PAYLOAD locators, we can only interrogate the payload while the request
+      // is being evaluated, via peekData()
+      LOGGER.debug("Checking authorization before in peekData.");
+
+      // This is a little ugly - peekData provides only a block of data at a time, so lets
+      // buffer it up until we think its complete.
+      ByteArrayOutputStream dataStream = new ByteArrayOutputStream();
+
+      // Note that peekData only gets called for requests that actually have data (like PUT and POST)
+
+      var peekReq = req.peekData(data -> {
+        LOGGER.debug("Authorization peekData invoked.");
+
+          // TODO: For now, we're going to assume JSON data, but might need to support other
+          // content types.
+        if (req.contentType().equals(MediaType.JSON)) {
+          try {
+            dataStream.write(data.array());
+            LOGGER.debug("Payload: {}", dataStream.toString());
+          } catch (IOException e) {
+            // IGNORE
+          }
+
+          // Unfortunately we don't appear to get a signal that we've got all the data, so we have to
+          // resort to attempting to parse the data whenever it _looks_ complete.
+          // TODO: try to optimize this using Jackson streaming or something else.
+          if (data.array()[data.array().length - 1] == '}') {
+            try {
+              Map<String, Object> payload = MAPPER.readValue(dataStream.toByteArray(), new TypeReference<>() {
+              });
+
+              payloadLocators.forEach(l -> resourceKeys.put(l.getType(), findPayloadValue(l.getKey(), payload)));
```

Comment:
probably better to break this into a separate method? `extractResoureKeysFromPayload`?

Reply:
I broke this out to a separate class. It can't just be a method it needs to keep state since the lambda gets called multiple times per request. But breaking it out to a class so it can encapsulate the state is a good idea.